### DISCUSSION TITLE
Flattened operator node not only at top-level

### DISCUSSION THREAD

**Discussion by bostick** [tree-sitter/tree-sitter#1456] [Q&A] [answered]
This is split off from a previous discussion:
https://github.com/tree-sitter/tree-sitter/discussions/1429

With the grammar from the previous discussion, the input `(a < b < c) < d < e;` parses as:
```
(program [0, 0] - [0, 20]
  (expression [0, 0] - [0, 19]
    (cmp_expr [0, 0] - [0, 19]
      base: (id [0, 1] - [0, 2])
      op: (compare_op [0, 3] - [0, 4])
      base: (id [0, 5] - [0, 6])
      op: (compare_op [0, 7] - [0, 8])
      expr: (id [0, 9] - [0, 10])
      op: (compare_op [0, 12] - [0, 13])
      base: (id [0, 14] - [0, 15])
      op: (compare_op [0, 16] - [0, 17])
      expr: (id [0, 18] - [0, 19]))))
```

e.g. the parens are lost.

It seems the answer grammar is "top-level only".

Is there a way to achieve "flattened operator" for anywhere inside of a grammar?

**Comment by ahelwer** [comment, marked as answer]
I don't think there is really a good way to do this. I tried using [negative dynamic precedence](https://stackoverflow.com/a/66932263/2852699) but it affected the rest of the grammar in a way that made it difficult to develop. I also couldn't think of an actual use-case for this. Any interpreter would traverse a nested left-associative structure pretty much the same as it would a flat structure. I suppose maybe it could be used to implement rainbow highlighting for operator nesting or something.

**Reply by leonerd** [reply]
There are many use-cases. Many programming languages now have chaining comparison operators, or other situations where you need flattened list precedence. It doesn't immediately appear to be possible; which is a shame. I would like to suggest adding a new kind of `prec` function `prec.list()` for "list precedence", so you could write this as:

```
cmp_expr: $ => prec.list(10,
    seq(
        field("base", $.expression),
        repeat1(seq(
            field("op", $.compare_op),
            field("expr", $.expression),
        )),
    )
),
```

This would have the effect of flattening subsequent expansions of itself, but only when they're immediate. It would allow you to implement `1 == 2 == 3` as a flat `cmp_expr` of three numbers, while still preserving the tree-like structure of both `(1 == 2) == 3` and `1 == (2 == 3)`.

I don't think it is possible to implement without this.

**Reply by ahelwer** [reply]
Well if you check on the stack overflow question I linked someone actually came up with an answer. I should update it as the answer to this question as well.

**Comment by bostick** [comment]
OK, thank you for the answer.

The use-case of `a < b < c` in the first discussion I think is a good example of flat operators. Some languages do parse `a < b < c` as a single boolean expression, and of course they may not only occur in top-level.

My personal use-case is related to parsing `a + b + c` and `a * b * c` for Wolfram Language.

It is common with machine-generated WL code that there may be `+` or `*` expressions that span hundreds of lines and thousands of operands.

These are parsed "flat" and it is a huge efficiency gain in terms of both memory and speed.

**Reply by kenkangxgwe** [reply]
Haven't touched TS before, but what if you write the infixes with `repeat1`?
```js
 infix: $ => choice(
      ...
      prec.left(PRECEDENCE_SEMI, seq($._expression, repeat1(';', $._expression))),
      ...
```
